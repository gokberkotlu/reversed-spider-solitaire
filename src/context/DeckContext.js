import React, { useState, createContext, useEffect } from "react";
import { cardSeries } from '../utils/CardSeries';

export const DeckContext = createContext();

export const DeckProvider = props => {

    // States
    const [gameCards, setGameCards] = useState([[],[],[],[],[],[],[],[],[],[]]); // Represents cards on table
    const [remainingCards, setRemainingCards] = useState([]); // Cards outside the game
    const [numberOfCompleted, setNumberOfCompleted] = useState(Array(0)); // Shows number of completed series
    const [newCardPacks, setNewCardPacks] = useState(Array(5)); // Shows number of new card packs
    const [isGameStarted, setIsGameStarted] = useState(false); // Set state when any card action happened
    const [statesLog, setStatesLog] = useState([]); // Log States to Undo the Game

    // States: Score Datas
    const [score, setScore] = useState(500); // Game score which starts 500
    const [highScore, setHighScore] = useState(+localStorage.getItem('highScore')); // Recorded highscore to store in localStorage

    const dealTheCards = (cards, numberOfCards) => {
        
        // Save Changes When the New Card Pack Added
        logGameChanges();

        if(cards.length >= 10) {
            let gameSet = [...gameCards];
            for(let i = 0; i < numberOfCards; i++) {
                gameSet[i%10].push(cards[i]);
            }
            setGameCards(gameSet);
            setRemainingCards(cards.splice(numberOfCards));

            return gameSet;
        }
    }

    // arrangeReversedCards arranges the reversed card at the start
    const arrangeReversedCards = (gameSet) => {
        gameSet.forEach(cards => {
            for(let i = 0; i < cards.length - 1; i++) {
                cards[i].isCardReverse = true;
            }
        })
    }

    // suffleTheCards shuffles eight decks of cards
    const suffleTheCards = () => {
        let deskNumber = 8;
        let newArray = [];
        for(let i = 0; i < deskNumber; i++) {
            newArray = newArray.concat(
                [...cardSeries.map((card) => {
                    return {
                        "cardNum": card,
                        "cardDeck": (i + 1),
                        "isCardReverse": false
                    }
                })]);
        }
        var currentIndex = newArray.length,
            randomIndex;
        // While there remain elements to shuffle...
        while (0 !== currentIndex) {
            // Pick a remaining element...
            randomIndex = Math.floor(Math.random() * currentIndex);
            currentIndex--;
    
            // And swap it with the current element.
            [newArray[currentIndex], newArray[randomIndex]] = [
                newArray[randomIndex], newArray[currentIndex]
            ];
        }
        return newArray;
    }

    // logGameChanges takes snapshots before all card actions and stores them inside statesLog
    const logGameChanges = () => {
        let addToLog = {
            "gameCardsBefore": JSON.parse(JSON.stringify(gameCards)),
            "remainingCardsBefore": JSON.parse(JSON.stringify(remainingCards)),
            "numberOfCompletedBefore": JSON.parse(JSON.stringify(numberOfCompleted)),
            "newCardPacksBefore": JSON.parse(JSON.stringify(newCardPacks))
        }
        setStatesLog(prevState => [...prevState, addToLog]);
    }

    // undoTheGame sets game states with last logged snapshot
    const undoTheGame = () => {
        let statesLogValue = [...statesLog];
        let statesLogValueLength = statesLogValue.length
        if(statesLogValueLength > 1) {
            let beforeStates = statesLogValue[statesLogValueLength - 1]; // Go back in the game by having "before value of last screen"
            setGameCards(beforeStates.gameCardsBefore);
            setRemainingCards(beforeStates.remainingCardsBefore);
            setNumberOfCompleted(beforeStates.numberOfCompletedBefore);
            setNewCardPacks(beforeStates.newCardPacksBefore);

            // Delete current states from log when went back in the game
            setStatesLog(statesLogValue.slice(0, statesLogValueLength - 1));

            // Decrease 1 point from score for each undo click
            setScore(prevState => prevState - 1);

            // Decrease 100 points from score if completed desk is removed by clicking undo button
            if(beforeStates.numberOfCompletedBefore.length < numberOfCompleted.length) {
                let scoreDiff = numberOfCompleted.length - beforeStates.numberOfCompletedBefore.length;
                setScore(prevState => prevState - 100 * scoreDiff);
            }
        }
    }

    useEffect(() => {
        // Shuffle cards, arrange reversed cards and deal them on the game table at start
        var allCards = suffleTheCards();
        var gameSet = dealTheCards(allCards, 54);
        arrangeReversedCards(gameSet);
    }, [])

    return (
        <DeckContext.Provider
        value={{
                "gameCards": [gameCards, setGameCards],
                "remainingCards": [remainingCards, setRemainingCards],
                "numberOfCompleted": [numberOfCompleted, setNumberOfCompleted],
                "newCardPacks": [newCardPacks, setNewCardPacks],
                "isGameStarted": [isGameStarted, setIsGameStarted],
                "suffleTheCards": suffleTheCards,
                "dealTheCards": dealTheCards,
                "arrangeReversedCards": arrangeReversedCards,
                "statesLog": [statesLog, setStatesLog],
                "logGameChanges": logGameChanges,
                "undoTheGame": undoTheGame,
                "score": [score, setScore],
                "highScore": [highScore, setHighScore]
            }}>
            {props.children}
        </DeckContext.Provider>
    );
}