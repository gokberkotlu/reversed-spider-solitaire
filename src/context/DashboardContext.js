import React, { useState, createContext, useContext, useEffect } from "react";
import { DeckContext } from "./DeckContext";

export const DashboardContext = createContext();

export const DashboardProvider = props => {

    // States
    const [counter, setCounter] = useState("00:00:00"); // Represents timer
    const [isGameFinished, setIsGameFinished] = useState(false); // According to this state, popup is shown or not
    const [resetBit, setResetBit] = useState(false); // It triggers the useEffect to reset the game
    const [decreaseInEveryTenSeconds, setDecreaseInEveryTenSeconds] = useState(0); // Decrease 2 points from score for each 10 sec

    // DeckContext
    const deckContextValue = useContext(DeckContext);

    const [, setGameCards] = deckContextValue.gameCards;
    const [, setRemainingCards] = deckContextValue.remainingCards;
    const [, setNumberOfCompleted] = deckContextValue.numberOfCompleted;
    const [, setNewCardPacks] = deckContextValue.newCardPacks;
    const [, setIsGameStarted] = deckContextValue.isGameStarted;
    const [, setStatesLog] = deckContextValue.statesLog;
    const suffleTheCards = deckContextValue.suffleTheCards;
    const dealTheCards = deckContextValue.dealTheCards;
    const arrangeReversedCards = deckContextValue.arrangeReversedCards;

    const [, setScore] = deckContextValue.score;

    const resetGame = () => {
        // Resetting
        setGameCards([[],[],[],[],[],[],[],[],[],[]]);
        setRemainingCards([]);
        setNewCardPacks(Array(5));
        setNumberOfCompleted(Array(0));
        setScore(500);
        setCounter("00:00:00");
        setIsGameFinished(false);
        setResetBit(true);
        setIsGameStarted(false);
        setStatesLog([]);
        setDecreaseInEveryTenSeconds(0);
    }

    useEffect(() => {// Deal cards again from scratch
        if(resetBit) {
            var allCards = suffleTheCards();
            var gameSet = dealTheCards(allCards, 54); // setGameCards and setRemainingCards included
            arrangeReversedCards(gameSet);
            setResetBit(false);
        }
    }, [resetBit]) // resetBit changes trigger the control of gameSet and remainingCards

    return (
        <DashboardContext.Provider
        value={{"counter": [counter, setCounter],
        "isGameFinished": [isGameFinished, setIsGameFinished],
        "resetGame": resetGame,
        "decreaseInEveryTenSeconds": [decreaseInEveryTenSeconds ,setDecreaseInEveryTenSeconds]}}>
            {props.children}
        </DashboardContext.Provider>
    );
}