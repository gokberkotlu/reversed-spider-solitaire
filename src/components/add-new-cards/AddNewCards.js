import CardBack from "../card/card-templates/card-back/CardBack";
import { useContext } from "react";
import { DeckContext } from "../../context/DeckContext";
import './add-new-cards.css'

const AddNewCards = () => {
    
    // DeckContext
    const deckContextValue = useContext(DeckContext);

    const dealTheCards = deckContextValue.dealTheCards;
    const [remainingCards,] = deckContextValue.remainingCards;
    const [newCardPacks, setNewCardPacks] = deckContextValue.newCardPacks;
    const [isGameStarted, setIsGameStarted] = deckContextValue.isGameStarted;
    const [, setScore] = deckContextValue.score;

    return (
        <div className="add-new-cards-container">
            {[...newCardPacks].map((val, index) => {
                if(index === newCardPacks.length - 1) {
                    return (
                        <div key={'pack-' + index}
                        className="add-new-cards"
                        onClick={() => {

                                // Deal new card pack
                                dealTheCards(remainingCards, 10);
                                let newCardPacksClone = [...newCardPacks];
                                newCardPacksClone.pop();
                                setNewCardPacks(newCardPacksClone);

                                // Start timer when first 10 card pack added
                                if(!isGameStarted) {
                                    setIsGameStarted(true);
                                }

                                // Decrease score by 1 when new card pack added
                                setScore(prevState => prevState - 1);
                            }}>
                            <CardBack />
                        </div>
                    )
                } else {
                    return (
                        <div key={'pack-' + index} className="add-new-cards">
                            <CardBack />
                        </div>
                    )
                }
            })}
        </div>
    );
}
 
export default AddNewCards;