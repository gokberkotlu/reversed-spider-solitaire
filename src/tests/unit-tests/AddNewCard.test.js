import renderer from "react-test-renderer";
import { render, fireEvent } from '@testing-library/react';
import { DeckProvider } from '../../context/DeckContext';
import AddNewCards from "../../components/add-new-cards/AddNewCards";

describe('Add new card packs', () => {
    it('should render AddNewCard component', () => {
        const tree = renderer.create(
            <DeckProvider>
                <AddNewCards />
            </DeckProvider>
        ).toJSON();
        expect(tree).toMatchSnapshot();
    })

    it('should render five elements which have "add-new-cards" class', () => {
        render(
            <DeckProvider>
                <AddNewCards />
            </DeckProvider>
        );
        expect(document.getElementsByClassName("add-new-cards").length).toBe(5);
    })

    it('should decrease the number of new card pack containers by one, when last of them is clicked', () => {
        render(
            <DeckProvider>
                <AddNewCards />
            </DeckProvider>
        );
        const addNewCardElements = document.querySelectorAll(".add-new-cards");
        const lastAddNewCardElement = addNewCardElements[addNewCardElements.length - 1];
        fireEvent.click(lastAddNewCardElement);
        expect(document.getElementsByClassName("add-new-cards").length).toBe(4);
    })
})