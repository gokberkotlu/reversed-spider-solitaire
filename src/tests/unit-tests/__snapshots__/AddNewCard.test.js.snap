// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Add new card packs should render AddNewCard component 1`] = `
<div
  className="add-new-cards-container"
>
  <div
    className="add-new-cards"
  >
    <div
      className="card-template card-spade card-back"
    />
  </div>
  <div
    className="add-new-cards"
  >
    <div
      className="card-template card-spade card-back"
    />
  </div>
  <div
    className="add-new-cards"
  >
    <div
      className="card-template card-spade card-back"
    />
  </div>
  <div
    className="add-new-cards"
  >
    <div
      className="card-template card-spade card-back"
    />
  </div>
  <div
    className="add-new-cards"
    onClick={[Function]}
  >
    <div
      className="card-template card-spade card-back"
    />
  </div>
</div>
`;
